### PR TITLE
Unify overheardIgp and igp

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- v3 uses the OverheadIgp as the default igp and removing an "inner" igp makes it simpler to deploy and interact with
- `InterchainGasPaymaster` has a `setDestinationGasOverheads` which sets it the overhead ands adds it directly to the specified gasAmount.
- Downstream changes to the `typescript/*` module in the igpConfig, checker, governor, deployer, etc

### Drive-by changes

None

### Related issues

- precursor to https://github.com/hyperlane-xyz/issues/issues/623
- related to https://github.com/hyperlane-xyz/issues/issues/637

### Backward compatibility

Yes

### Testing

Fuzz

### CODE REVIEW

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol [hunk 1]
```diff
@@ -54,6 +57,9 @@ contract InterchainGasPaymaster is
     /// @notice Keyed by remote domain, the gas oracle to use for the domain.
     mapping(uint32 => IGasOracle) public gasOracles;
 
+    /// @notice Destination domain => overhead gas amount on that domain.
+    mapping(uint32 => uint256) public destinationGasOverhead;
+
```
Comment (yorhodes) [inline solidity/contracts/hooks/igp/InterchainGasPaymaster.sol:62]: should these be a single mapping?

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol [hunk 2]
```diff
@@ -159,18 +195,18 @@ contract InterchainGasPaymaster is
...
+            _requiredGasAmount,
             _requiredPayment
         );
     }
 
     /**
      * @notice Quotes the amount of native tokens to pay for interchain gas.
      * @param _destinationDomain The domain of the message's destination chain.
-     * @param _gasAmount The amount of destination gas to pay for.
+     * @param _totalGasAmount The amount of destination gas to pay for.
```
Comment (yorhodes) [inline solidity/contracts/hooks/igp/InterchainGasPaymaster.sol:206]: this rename doesn't add clarity
adjust description accordingly

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol [hunk 3]
```diff
@@ -218,6 +254,22 @@ contract InterchainGasPaymaster is
         return _gasOracle.getExchangeRateAndGasPrice(_destinationDomain);
     }
 
+    /**
+     * @notice Returns the stored destinationGasOverhead added to the _gasAmount.
+     * @dev If there is no stored destinationGasOverhead, 0 is used.
```
Comment (yorhodes) [inline solidity/contracts/hooks/igp/InterchainGasPaymaster.sol:259]: should we have an immutable default? or revert?

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol [hunk 4]
```diff
@@ -218,6 +254,22 @@ contract InterchainGasPaymaster is
...
+     * @param _destinationDomain The domain of the message's destination chain.
+     * @param _gasAmount The amount of destination gas to pay for. This should not
+     * consider any gas that is accounted for in the stored destinationGasOverhead.
+     * @return The stored destinationGasOverhead added to the _gasAmount.
```
Comment (yorhodes) [inline solidity/contracts/hooks/igp/InterchainGasPaymaster.sol:263]: "should not consider any gas that is accounted for in the stored overhead" is not very clear
can you make the distinction more obvious? eg system overhead vs application overhead (ISM verification or message handling)

File: solidity/test/igps/InterchainGasPaymaster.t.sol [hunk 5]
```diff
@@ -24,14 +24,16 @@ contract InterchainGasPaymasterTest is Test {
     uint32 constant testOriginDomain = 22222;
     uint32 constant testDestinationDomain = 11111;
     uint256 constant testGasAmount = 300000;
+    uint256 constant testGasOverhead = 123000;
+    uint256 constant DEFAULT_GAS_USAGE = 69_420;
```
Comment (yorhodes) [inline solidity/test/igps/InterchainGasPaymaster.t.sol:28]: nit: change this please

File: solidity/test/igps/InterchainGasPaymaster.t.sol [hunk 6]
```diff
@@ -72,6 +73,63 @@ contract InterchainGasPaymasterTest is Test {
...
+    }
+
+    function testSetDestinationGasAmounts() public {
+        InterchainGasPaymaster.DomainConfig[]
+            memory configs = new InterchainGasPaymaster.DomainConfig[](2);
+        configs[0] = InterchainGasPaymaster.DomainConfig(
+            testDestinationDomain,
+            testGasOverhead
+        );
+        configs[1] = InterchainGasPaymaster.DomainConfig(4321, 432100);
```
Comment (yorhodes) [inline solidity/test/igps/InterchainGasPaymaster.t.sol:99]: nit: fuzz

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/contracts/hooks/igp/InterchainGasPaymaster.sol:62]: might even fit into a single slot

File: typescript/infra/config/environments/mainnet2/igp.ts [hunk 7]
```diff
@@ -24,24 +24,21 @@ function getGasOracles(local: MainnetChains) {
...
+        remote,
+        multisigIsmVerificationCost(
+          defaultMultisigIsmConfigs[remote].threshold,
+          defaultMultisigIsmConfigs[remote].validators.length,
+        ),
+      ]),
+    ),
+    upgrade: core[chain].upgrade,
+  };
+});
```
Comment (yorhodes) [inline typescript/infra/config/environments/mainnet2/igp.ts:44]: nit: why is the formatting different?

File: typescript/infra/config/environments/test/igp.ts [hunk 8]
```diff
@@ -19,23 +19,20 @@ function getGasOracles(local: TestChains) {
...
+      exclude(chain, chainNames).map((remote) => [
+        remote,
+        multisigIsmVerificationCost(
+          multisigIsm[remote].threshold,
+          multisigIsm[remote].validators.length,
+        ),
+      ]),
+    ),
+  };
+});
```
Comment (yorhodes) [inline typescript/infra/config/environments/test/igp.ts:38]: ditto

File: typescript/infra/config/environments/testnet3/igp.ts [hunk 9]
```diff
@@ -19,23 +19,20 @@ function getGasOracles(local: TestnetChains) {
...
+      exclude(chain, supportedChainNames).map((remote) => [
+        remote,
+        multisigIsmVerificationCost(
+          defaultMultisigIsmConfigs[remote].threshold,
+          defaultMultisigIsmConfigs[remote].validators.length,
+        ),
+      ]),
+    ),
+  };
+});
```
Comment (yorhodes) [inline typescript/infra/config/environments/testnet3/igp.ts:38]: ditto

File: typescript/infra/scripts/utils.ts [hunk 10]
```diff
@@ -340,7 +340,7 @@ export async function getRouterConfig(
         ? await multiProvider.getSignerAddress(chain)
         : owners[chain],
       mailbox: core.getContracts(chain).mailbox.address,
-      hook: igp.getContracts(chain).defaultIsmInterchainGasPaymaster.address,
+      hook: igp.getContracts(chain).interchainGasPaymaster.address,
```
Comment (yorhodes) [inline typescript/infra/scripts/utils.ts:343]: did you remove other usage of this?

File: typescript/infra/src/govern/HyperlaneIgpGovernor.ts [hunk 11]
```diff
@@ -84,15 +84,14 @@ export class HyperlaneIgpGovernor extends HyperlaneAppGovernor<
...
-              gasOverhead: gasOverhead,
-            } as OverheadIgp.DomainConfigStruct),
-        );
+        const configs: InterchainGasPaymaster.DomainConfigStruct[] =
+          Object.entries(violation.expected).map(
+            ([remote, gasOverhead]) =>
+              ({
+                domain: this.checker.multiProvider.getDomainId(remote),
+                gasOverhead: gasOverhead,
+              } as InterchainGasPaymaster.DomainConfigStruct),
```
Comment (yorhodes) [inline typescript/infra/src/govern/HyperlaneIgpGovernor.ts:93]: why is casting necessary?

File: typescript/sdk/src/gas/HyperlaneIgpChecker.ts [hunk 12]
```diff
@@ -72,8 +72,8 @@ export class HyperlaneIgpChecker extends HyperlaneAppChecker<
 
     await this.checkBytecode(
       chain,
-      'OverheadIGP',
-      contracts.defaultIsmInterchainGasPaymaster.address,
+      'InterchainGasPaymaster',
+      contracts.interchainGasPaymaster.address,
       [BytecodeHash.OVERHEAD_IGP_BYTECODE_HASH],
```
Comment (yorhodes) [inline typescript/sdk/src/gas/HyperlaneIgpChecker.ts:77]: we probably need to change this? cc @nambrot

File: solidity/test/igps/InterchainGasPaymaster.t.sol (hunk 5 above)
Comment (aroralanuk) [reply solidity/test/igps/InterchainGasPaymaster.t.sol:28]: in the test or the contract?

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol (hunk 4 above)
Comment (aroralanuk) [reply solidity/contracts/hooks/igp/InterchainGasPaymaster.sol:263]: this is @tkporter's comment.  The contract @dev comment clarifies it. I've added the "application-sepcific" part.

File: typescript/infra/src/govern/HyperlaneIgpGovernor.ts (hunk 11 above)
Comment (aroralanuk) [reply typescript/infra/src/govern/HyperlaneIgpGovernor.ts:93]: fixed

File: typescript/infra/config/environments/mainnet2/igp.ts (hunk 7 above)
Comment (aroralanuk) [reply typescript/infra/config/environments/mainnet2/igp.ts:44]: IgpConfig is shorter than OverheadIgpConfig

File: typescript/infra/config/environments/test/igp.ts (hunk 8 above)
Comment (aroralanuk) [reply typescript/infra/config/environments/test/igp.ts:38]: IgpConfig is shorter than OverheadIgpConfig

File: typescript/infra/config/environments/testnet3/igp.ts (hunk 9 above)
Comment (aroralanuk) [reply typescript/infra/config/environments/testnet3/igp.ts:38]: IgpConfig is shorter than OverheadIgpConfig

File: typescript/infra/scripts/utils.ts (hunk 10 above)
Comment (aroralanuk) [reply typescript/infra/scripts/utils.ts:343]: defaultIsmInterchainGasPaymaster is removed from all places except the config files.

File: typescript/sdk/src/gas/HyperlaneIgpChecker.ts (hunk 12 above)
Comment (aroralanuk) [reply typescript/sdk/src/gas/HyperlaneIgpChecker.ts:77]: cc @nambrot

File: solidity/contracts/hooks/igp/InterchainGasPaymaster.sol (hunk 3 above)
Comment (aroralanuk) [reply solidity/contracts/hooks/igp/InterchainGasPaymaster.sol]: added more context

File: solidity/test/igps/InterchainGasPaymaster.t.sol (hunk 6 above)
Comment (aroralanuk) [reply solidity/test/igps/InterchainGasPaymaster.t.sol]: fuzzed

**Review by yorhodes — APPROVED** [review]

File: typescript/sdk/src/gas/HyperlaneIgpChecker.ts (hunk 12 above)
Comment (aroralanuk) [reply typescript/sdk/src/gas/HyperlaneIgpChecker.ts:77]: refer to https://github.com/hyperlane-xyz/issues/issues/642

**Pull request merged**